### ISSUE TITLE
Validate the shape of the input vector during inference

### ISSUE THREAD

**Issue by batata004** [RubixML/ML#90]
After training my model, I tried to predict a single sample. I tried:

`$prediction = $estimator->predictSample([0,0,1,0,1,0,1]);`

And it showed me the prediction. Then I realized I had forgot to include the `label` in that line, so I added the label, and it ended up like this:

`$prediction = $estimator->predictSample([0,0,1,0,1,0,1,"sim"]);`

And it showed me again the prediction. Then I got confused? Some of those 2 lines should not work because I am input a data that has different shape that the training. Then I tried adding more 0,1 items in the `predictSample` and it always predicts. No error shown!

This makes that function `predictSample` unreliable since it should only accepts the same number of columns than the original training dataset. Also, do I need to provide the label or not? Will it make any difference? I think the label should not be provided since it's a prediction, not a training and the label should have no effect on the output of the prediction!

**Comment by andrewdalpino** [comment]
Hi Gil!

You would not include a label during inference. Labels are only necessary for supervised training and cross-validation. In addition, if a label was required, it would not be mixed in with the features. This as well as a note to maintain the order of features is all documented.

Here's a link to the documentation https://docs.rubixml.com/en/latest/inference.html

**Comment by batata004** [comment]
@andrewdalpino thank you, now I understand that! But did you notice that when I provide an array with less or more features/columns at `predictSample` than it had originally during training, no error is shown? So if I provide less features, what is happenig under the hood? And if I provide more? I think that if the number of features is different at `predictSample` it should throw and error and alert the user to fix the problem. It should not silently accept the error made by the user (that certainly was no intentional) and keep processing and if nothing was wrong! Right?

**Comment by andrewdalpino** [comment]
I see what you're saying @batata004 and yes I think it would be nice to know when the number of input features does not match the number used to train the learner and I can see how that would have benefited you in this circumstance. Let's see if anyone else runs into this type of issue - this is the first case so far. I'm thinking we could implement a 'DatasetHasDimensionality' specification to be used in the `predict()`, `proba()`, and `rank()` methods of every learner.

Note that this would not validate the order of the features, only the number of them. What are your thoughts?

**Comment by batata004** [comment]
@andrewdalpino Really great! Indeed, it would only validate the number of columns/features and not the order. But I think I may be missing something really important in this ML subject: does it make any sense running a prediction without providing the same number of features/columns that you provided when you trained the model? I mean, if you trained your model providing always 20 features, will that model be any good if I try to predict something providing only 3 features to that same model? Or providing 100 features? If the model was trained using 20 features, in a very specific order, would the model have any good prediction if the person provides only 3 features or if the features are provided in wrong order? What does the ML model does to replace the other features that were not provided or what does the ML model does if the person provides more features than during training?

Maybe I am missing something really important here, cause to me it make little sense running a prediction providing less or more features than you trained your model before! Right? I am possibly wrong in this, and I would be very glad if you could explain me what I am missing!

**Comment by andrewdalpino** [comment]
Good questions @batata004!

The internal model depends both on the number and order of features. The same features used to train the model are the ones that should be used to derive predictions from it. It may help to think of the values of each feature as a coordinate in some n-dimensional vector space. Thus, swapping coordinates will (most likely) change the direction of the vector. In other words, it would represent something else. Also, dropping a column can be thought of as effectively removing a degree of freedom. It would be like you or I living and operating in a 2-dimensional world rather than a 3-dimensional one and as bad as that sounds it's equally bad for a learner trying to make sense of the data its being fed. Does that make sense?

tldr; The number and order of features should be the same for both training and inference.

I'm wondering if there was a better way we could clarify this in the docs. Any suggestions?

**Comment by batata004** [comment]
@andrewdalpino everything you said makes total sense to me! And I think what you said is very intuitive and I dont know if it would be necessary to say in the documentation that the number of features, when predicting a sample, should be the same number of features when training the model - and also the order should be the same. I think that the code itself should alert the user when trying to predict a sample with a different number of features than the ones used in training. Then the user would realize that most probably made an unintentional error! But of course, I am pretty new to this ML stuff and surely I dont have much experience (if any) to provide any advice!

Anyway, I am closing this issue because clearly I was wrong, it's not a bug at all, and it's up to most experient people discuss if the user should be told when his sample has a different number of features than when training! 

I appreciate a lot your attention to clarifying things to newbies like me :)

**Comment by andrewdalpino** [comment]
I'm reopening this to implement input dimensionality checks for estimators where appropriate

Thanks again @batata004 for your input!

**Comment by andrewdalpino** [comment]
Implemented in https://github.com/RubixML/RubixML/commit/d1b53d363cd1cbd241a85a18fa3f0c2de5bb5484
